includes/class-database-cache.php

```diff
@@ -244,6 +310,17 @@ private function should_refresh_cache( string $key, $cache_contents, callable $v
 			return true;
 		}
 
+		// Do not refresh if we had DB cache write errors for this key and we have a valid in-memory cache.
+		// For the remainder of the request, we will what we have in the in-memory cache,
+		// if the cache is not explicitly cleared or forced to refresh.
+		if ( isset( $this->db_cache_write_errors[ $key ] ) &&
+			$this->db_cache_write_errors[ $key ] > 0 &&
```

Comment:
This suggestion sounds sane to me. I cannot imagine a performance hit if we always use the in-memory cache fallback.

Reply:
@marcinbot, I searched long and hard for a way to forego the use of `db_cache_write_errors`, but I haven't found one.

Now that we got rid of the cache size management let me explain the current logic; maybe you spot something I missed:
- we always initialize the in-memory cache when reading (or writing) the cache.
- we always use the data in the in-memory cache (there will be only one `get_option` call per key, per request), unless the cache entry is deleted or we are forced to refresh _mid-request_
- if a cache entry needs to be refreshed (missing, expired, invalid, etc), we need to know if there were prior DB write failures:
  - if there were no failures, we will proceed as usual and refresh the cached data
  - if there were failures, we don't try the refresh and instead use what we have in the in-memory cache (data fetched from our platform but not written to the DB). Without knowing of the previous DB write failures, we cannot stop _continuously_ refreshing the data on every call to `get_or_add`.

I don't think we have the option of making the in-memory cache sticky per an entire request because of the `force_refresh` logic.

Does this make sense?

Reply:
Oh, man! It was staring me in the face and I didn't see it - I got too caught up with errored responses, but they are not retried until the cache expires.

I removed the DB write errors handling, and it works like a charm! Tested it with correct platform responses, empty array, and errors (rate limited), all with simulated DB write failures.

Thank you for not backing off so easily, Marcin! Now, we have a truly elegant solution. Excellent!